core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+    private Map<String, String> properties = Maps.newHashMap();
+    private List<MetadataUpdate> changes = Lists.newArrayList();
 
-    int versionHistorySizeToKeep =
-        PropertyUtil.propertyAsInt(
-            properties(),
-            ViewProperties.VERSION_HISTORY_SIZE,
-            ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
-
-    if (versionHistorySizeToKeep <= 0) {
-      LOG.warn(
-          "{} must be positive but was {}",
-          ViewProperties.VERSION_HISTORY_SIZE,
-          versionHistorySizeToKeep);
-    } else if (versions().size() > versionHistorySizeToKeep) {
-      List<ViewVersion> versions =
-          versions().subList(versions().size() - versionHistorySizeToKeep, versions().size());
-      List<ViewHistoryEntry> history =
-          history().subList(history().size() - versionHistorySizeToKeep, history().size());
-      return ImmutableViewMetadata.builder().from(this).versions(versions).history(history).build();
+    private Builder() {}
+
+    private Builder(ViewMetadata base) {
+      this.formatVersion = base.formatVersion();
+      this.location = base.location();
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.currentVersionId = base.currentVersionId();
+      this.versions = Lists.newArrayList(base.versions());
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
     }
 
-    return this;
+    public Builder upgradeFormatVersion(int newFormatVersion) {
+      Preconditions.checkArgument(
+          newFormatVersion >= this.formatVersion,
+          "Cannot downgrade v%s view to v%s",
+          formatVersion,
+          newFormatVersion);
+
+      if (this.formatVersion == newFormatVersion) {
+        return this;
+      }
+
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      Preconditions.checkArgument(null != newLocation, "Invalid location: null");
+      if (null != this.location && this.location.equals(newLocation)) {
+        return this;
+      }
+
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      this.currentVersionId = newVersionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder setSchemas(Iterable<Schema> schemasToAdd) {
```

Comment:
This has the same problem that I pointed out about `setVersions` here: https://github.com/apache/iceberg/pull/8147#pullrequestreview-1575884865. There's no need to have this method, so let's remove it.

Reply:
done